HW2/P4/P4.txt

```diff
@@ -0,0 +1,13 @@
+(Plot: num_threads_result.png)
+
+In this problem, the original image is divided into "stripes" where every thread can perform median filtering on one set of stripes. For example, in the case with 4 threads, thread 1 will work on 0,4,8,... and thread 2 will work on 1,5,9,... After sufficient number of iterations, the parallelized output filtered image will be the same as that done by serial numpy median code. 
+
+However, the transformation under this parallel case should use extra caution because when performing median filtering, one thread may be transforming certain line another thread is currently reading. Therefore, I used events to avoid such issue. I initialized the events matrix that will store all the event states of threads in multiple iterations. When one thread is transforming the lines, it "locks" (set their event to wait) both the line above it and the line below it. When finished processing (no collision issue any more), the event will be set.
```

Comment:
Your explanation forgets that the prior iterations of the adjacent threads must be complete before the current iteration of a thread. However, your code accounts for this.